File: dsa-problems/Queues/solutions/python/QUE-005-lab-printer-reversal.py
from typing import List
from collections import deque
import sys

def reverse_first_k(values: List[int], k: int) -> List[int]:
    queue = deque(values)
    stack = []
    
    # 1. Dequeue first K and push to stack
    for _ in range(k):
        stack.append(queue.popleft())
        
    # 2. Pop from stack and enqueue
    while stack:
        queue.append(stack.pop())
        
    # 3. Rotate remaining N-K elements
    n = len(values)
    for _ in range(n - k):
        queue.append(queue.popleft())
        
    return list(queue)

def main():
    input_data = sys.stdin.read().split()
    if not input_data:
        return

    iterator = iter(input_data)
    try:
        n = int(next(iterator))
        k = int(next(iterator))
        values = [int(next(iterator)) for _ in range(n)]

        result = reverse_first_k(values, k)
        print(" ".join(map(str, result)))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/Queues/solutions/javascript/QUE-005-lab-printer-reversal.js
const readline = require("readline");

class Solution {
  reverseFirstK(values, k) {
    // Using array methods to simulate Queue/Stack for clarity
    const queue = [...values]; // Copy
    const stack = [];
    
    // 1. Dequeue first K
    for (let i = 0; i < k; i++) {
      stack.push(queue.shift());
    }
    
    // 2. Pop and Enqueue
    while (stack.length > 0) {
      queue.push(stack.pop());
    }
    
    // 3. Rotate remaining N-K
    const n = values.length;
    for (let i = 0; i < n - k; i++) {
      queue.push(queue.shift());
    }
    
    return queue;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/).filter(x => x !== "")));
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  const k = parseInt(data[idx++], 10);
  const values = [];
  for (let i = 0; i < n; i++) {
    values.push(parseInt(data[idx++], 10));
  }

  const solution = new Solution();
  const result = solution.reverseFirstK(values, k);
  console.log(result.join(" "));
});
